pkg/tasks/runner.go

```diff
@@ -33,25 +34,78 @@ func RunAll(cfg *config.Config, proj *project.Project, ui *termui.UI) (success b
 	}
 
 	for _, task := range taskList {
-		ctx.ui.TaskHeader(task.name(), task.header())
+		if t, ok := task.(taskWithPreRunValidation); ok {
+			err = t.preRunValidation(ctx)
+			if err != nil {
+				ctx.ui.TaskError(err)
+				return false, nil
+			}
+		}
+	}
 
-		err = task.perform(ctx)
+	for _, task := range taskList {
+		ctx.ui.TaskHeader(task.name(), task.header())
+		err = runTask(ctx, task)
 		if err != nil {
 			ctx.ui.TaskError(err)
-			return false, nil
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+func runTask(ctx *context, task Task) (err error) {
+	if t, ok := task.(taskWithPerform); ok {
+		err = t.perform(ctx)
+		if err != nil {
+			return err
 		}
+	}
 
-		err = activateFeature(ctx, task)
+	for _, action := range task.actions(ctx) {
+		err = runAction(ctx, action)
 		if err != nil {
-			ctx.ui.TaskError(err)
-			return false, nil
+			return err
 		}
 	}
 
-	return true, nil
+	err = activateFeature(ctx, task)
+	return err
+}
+
+func runAction(ctx *context, action taskAction) error {
+	desc := action.description()
+
+	needed, err := action.needed(ctx)
+	if err != nil {
+		return fmt.Errorf("The task action (%s) failed to detect whether it need to run: %s", desc, err)
+	}
+
+	if desc != "" {
+		ctx.ui.TaskActionHeader(desc)
+	}
```

Comment:
I'm still not sure to understand why some actions has a description and some others don't

Reply:
The idea is some like... exactly this. 
Some actions have a description (so we display it), some don't (eg: there is only one action, and no relevant information to display)